File: 2020/src/day11.ml
open Lib

let read_cell l (x, y) = try List.(get y l |> get x) with _ -> '.'

let inp () = read_file "day11" string_2_char_list

let neigh pos l =
  let open List in
  init 8 (fun elt ->
      let elt = elt + if elt >= 4 then 1 else 0 in
      pos ++ ((elt / 3 mod 3) - 1, (elt mod 3) - 1))

let rec calc l =
  let l' =
    List.(
      map
        (fun l' ->
          map
            (fun elt ->
              let x = neigh elt l |> map (read_cell l) |> count '#' in
              match read_cell l elt with
              | 'L' when x = 0 -> '#'
              | '#' when x > 3 -> 'L'
              | z -> z)
            l')
        (init (length l) (fun y -> init (length (hd l)) (fun x -> (x, y)))))
  in
  if l' = l then l else calc l'

let day11_1 () = inp () |> calc |> List.flatten |> List.count '#'

let first_pos l x step =
  let rec aux ((x', y') as pos') =
    try
      let res = List.(get y' l |> get x') in
      if res <> '.' then res else aux (step ++ pos')
    with _ -> '.'
  in
  aux (step ++ x)

let neigh' () =
  let open List in
  init 8 (fun elt ->
      let elt = elt + if elt >= 4 then 1 else 0 in
      ((elt / 3 mod 3) - 1, (elt mod 3) - 1))

let rec calc' l =
  let l' =
    List.(
      map
        (fun l' ->
          map
            (fun elt ->
              let x = neigh' () |> map (first_pos l elt) |> count '#' in
              match read_cell l elt with
              | 'L' when x = 0 -> '#'
              | '#' when x > 4 -> 'L'
              | z -> z)
            l')
        (init (length l) (fun y -> init (length (hd l)) (fun x -> (x, y)))))
  in
  if l' = l then l else calc' l'

let day11_2 () = inp () |> calc' |> List.flatten |> List.count '#'

let _ = day11_2 () |> print_int

File: 2021/src/day11.py
f = './inp/day11.txt'


def matrix():
    with open(f) as F:
        return [[int(i) for i in line.strip()] for line in F.readlines()]


def neigh(M, x, y):
    return [(a, b) for (a, b) in [(x-1, y-1), (x-1, y), (x-1, y+1), (x, y-1), (x, y+1), (x+1, y-1), (x+1, y), (x+1, y+1)] if a >= 0 and b >= 0 and a < len(M[0]) and b < len(M)]


def increase_by_one(M):
    for line in M:
        for x in range(len(line)):
            line[x] += 1


def flash_mat(M, acc=0):
    mod = False
    for y, line in enumerate(M):
        for x, elt in enumerate(line):
            if elt >= 10:
                acc += 1
                N = neigh(M, x, y)
                for a, b in N:
                    M[b][a] += M[b][a] != 0
                    mod = True
                line[x] = 0
    return flash_mat(M, acc) if mod else acc


def day11_1():
    M = matrix()
    acc = 0
    for _ in range(100):
        increase_by_one(M)
        acc += flash_mat(M)
    return acc


def all_flash(M):
    res = flash_mat(M)
    return res == len(M) * len(M[0])


def day11_2():
    acc = 0
    M = matrix()
    while True:
        acc += 1
        increase_by_one(M)
        if all_flash(M):
            return acc


print(day11_1())
print(day11_2())
